src/core/geometry/qgsgeometry.cpp

```diff
@@ -1772,6 +1779,37 @@ QVector<QgsGeometry> QgsGeometry::coerceToType( const Qgis::WkbType type, double
     newGeom = QgsGeometry( std::move( mp ) );
   }
 
+  //(Multi)Polygon to PolyhedralSurface
+  if ( QgsWkbTypes::flatType( type ) == Qgis::WkbType::PolyhedralSurface &&
+       QgsWkbTypes::flatType( QgsWkbTypes::singleType( newGeom.wkbType() ) ) == Qgis::WkbType::Polygon )
+  {
+    std::unique_ptr< QgsPolyhedralSurface > polySurface = std::make_unique< QgsPolyhedralSurface >();
+    const QgsGeometry source = newGeom;
+    for ( auto part = source.const_parts_begin(); part != source.const_parts_end(); ++part )
+    {
+      std::unique_ptr< QgsAbstractGeometry > partAbstractGeometry( ( *part )->clone() );
```

Comment:
I may miss something, but the clone of the part doesn't seem necessary, since you don't actually add it to another component, but cast it as a polygon, and clone that polygone.

Reply:
You're right. Fixed.